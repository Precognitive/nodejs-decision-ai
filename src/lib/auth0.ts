// Auth0 Types, Interfaces and other structures
// ---------------------------------------------------


/**
 * @description Auth0 authentication protocol potential values
 * @link https://auth0.com/docs/rules/references/context-object
 */
const enum ContextProtocol {
  OidcBasicProfile = 'oidc-basic-profile',
  OidcImplicitProfile = 'oidc-implicit-profile',
  OAuth2ResourceOwner = 'oauth2-resource-owner',
  OAuth2ResourceOwnerJwtBearer = 'oauth2-resource-owner-jwt-bearer',
  OAuth2Password = 'oauth2-password',
  OAuth2RefreshToken = 'oauth2-refresh-token',
  SAMLP = 'samlp',
  WSFed = 'wsfed',
  WSTrustUsernameMixed = 'wstrust-usernamemixed',
  Delegation = 'delegation',
  RedirectCallback = 'redirect-callback',
}

/**
 * @description Auth0 Context Object, passed into rules
 * @link https://auth0.com/docs/rules/references/context-object
 */
interface Context {
  clientID: string,
  protocol: ContextProtocol,
  sessionID: string,
  request: {
    hostname: string,
    ip: string,
    query: string,
    userAgent: string,
    geoip: {
      city_name: string,
      continent_code: string,
      country_code3: string,
      country_code: string,
      country_name: string,
      latitude: string,
      longitude: string,
      time_zone: string,
    }
  }
}

/**
 * @description Auth0 User Object, passed into rules
 * @link https://auth0.com/docs/rules/references/user-object
 */
interface User {
  app_metadata: object,
  created_at: Date,
  email: string,
  last_ip: string,
  last_login: Date,
  logins_count: number,
  last_password_reset: Date,
  name: string,
  password_set_date: Date,
  updated_at: Date,
  username: string,
  user_id: string,
  user_metadata: object
}

/**
 * @description Auth0 Callback passed into rules
 * @link: https://auth0.com/docs/rules#syntax
 */
interface Callback {
  (err: null | Error, user: User, context: Context): void
}

/**
 * @description Auth0 Rule Interface
 * @link https://auth0.com/docs/rules
 */
interface Rule {
  (user: User, context: Context, callback: Callback): void
}

export { Context, ContextProtocol, User, Callback, Rule };
