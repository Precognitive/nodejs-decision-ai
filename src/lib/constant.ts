export const SDK_NAME = 'Cognition';
