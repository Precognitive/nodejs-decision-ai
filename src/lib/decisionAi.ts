// Cognition SDK
// ---------------------------------------------------

const enum ApiVersion {
  v1 = 'v1'
}

const enum DecisionStatus {
  allow = 'allow',
  review = 'review',
  reject = 'reject'
}

const enum Channel {
  web = 'web',
  desktop = 'desktop',
  app = 'app',
}

const enum LoginStatus {
  success = 'success',
  failure = 'failure'
}

const enum AuthenticationType {
  client_storage = 'client_storage',
  password = 'password',
  two_factor = 'two_factor',
  single_sign_on = 'single_sign_on',
  social_sign_on = 'social_sign_on',
  key = 'key'
}

interface CognitionResponse {
  score: number,
  confidence: number,
  decision: DecisionStatus,
  tokenId: string,
  signals: Array<string>
}

interface CognitionInput extends CognitionRequestOverrides {
  eventId?: string;
  dateTime?: Date;
  ipAddress: string;
  login: {
    userId: string;
    channel: Channel;
    usedCaptcha: boolean;
    usedRememberMe?: boolean;
    authenticationType?: AuthenticationType | null;
    status: LoginStatus;
    passwordUpdateTime: Date;
    userNameUpdateTime?: Date;
  };
}
type CognitionRequestOverrides = Readonly<CognitionInput>;

interface CognitionRequest extends CognitionInput {
  apiKey: string;
  dateTime: Date;
}

export {
  ApiVersion,
  DecisionStatus,
  Channel,
  LoginStatus,
  AuthenticationType,
  CognitionInput,
  CognitionResponse,
  CognitionRequest,
  CognitionRequestOverrides
};
