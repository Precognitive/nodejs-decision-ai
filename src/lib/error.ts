import { SDK_NAME } from './constant';

class DecisionError extends Error {
  public readonly isFraudulent: boolean;

  constructor(isFraudulent = true) {
    super('Precognitive: Reject Authentication');
    this.isFraudulent = isFraudulent;
  }
}

class HttpError extends Error {
  public readonly statusCode: number;
  public readonly response?: any;
  public readonly body?: any;

  constructor(statusCode: number, response?: any, body?: any) {
    super(`${SDK_NAME} - HTTP Error [${statusCode}]`);

    this.statusCode = statusCode;
    this.response = response;
    this.body = body;
  }
}

export {DecisionError, HttpError};
