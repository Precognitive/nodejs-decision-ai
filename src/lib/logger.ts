import { SDK_NAME } from './constant';

const enum LogLevel {
  DEBUG = 4,
  INFO = 3,
  WARN = 2,
  ERROR = 1,
  NONE = 0
}

class ConsoleLogger implements Logger {
  private readonly logLevel: LogLevel;

  constructor(logLevel: LogLevel) {
    this.logLevel = logLevel;
  }

  public debug(...args: any) {
    if (this.logLevel >= LogLevel.DEBUG) {
      console.debug(`${SDK_NAME} DEBUG:`, ...args);
    }
  }

  public info(...args: any) {
    if (this.logLevel >= LogLevel.INFO) {
      console.info(`${SDK_NAME} INFO:`, ...args);
    }
  }

  public warn(...args: any) {
    if (this.logLevel >= LogLevel.WARN) {
      console.warn(`${SDK_NAME} WARN:`, ...args);
    }
  }

  public error(...args: any) {
    if (this.logLevel >= LogLevel.ERROR) {
      console.error(`${SDK_NAME} ERROR:`, ...args);
    }
  }
}

interface Logger {
  debug(...args: any): void;
  info(...args: any): void;
  warn(...args: any): void;
  error(...args: any): void;
}

export {ConsoleLogger, LogLevel, Logger};
