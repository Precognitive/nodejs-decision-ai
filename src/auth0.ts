import _ from 'lodash';
import process from 'process';
import { Login } from './login';
import {
  User,
  Context,
  Callback
} from './lib/auth0';
import {
  ApiVersion,
  Channel,
  LoginStatus,
  AuthenticationType,
  CognitionResponse,
  CognitionRequestOverrides,
  DecisionStatus,
  CognitionInput
} from './lib/decisionAi';
import { ConsoleLogger, Logger, LogLevel } from './lib/logger';

interface ConstructorOptions {
  apiKey: string,
  version: ApiVersion,
  auth: {
    userName: string,
    password: string
  },
  getUserId?: {
    (user: User, context: Context): string
  },
  timeout?: number,
  logger?: Logger,
  logLevel?: LogLevel
}

interface DecisionOptions {
  defaultResponse?: DecisionStatus;
  overrides?: CognitionRequestOverrides;
  timeout?: number;
  privacyMode?: boolean;
}

class Auth0 {
  private readonly _base: Login;
  private readonly _getUserIdOption?: (user: User, context: Context) => string;
  private readonly _logger: Logger;

  constructor (options: ConstructorOptions) {
    if (options.logger) {
      this._logger = options.logger;
    } else {
      this._logger = new ConsoleLogger(options.logLevel || LogLevel.NONE);
    }

    this._getUserIdOption = options.getUserId;

    this._base = new Login({
      ...options,
      logger: this._logger
    });
  }

  /**
   * @description Sends login data for scoring and returns the result. `Auth0.isGoodLogin` may be used to determine if the login
   * is good. If there is a network outage, auth failure, or any other ip/tcp/http level error, it WILL return a rejected promise.
   */
  public async decision (user: User, context: Context, options: DecisionOptions = {}): Promise<CognitionResponse> {
    const reqBody = this._buildBody(user, context, options);
    return this._base.decision(reqBody, options);
  }

  /**
   * @description Sends login data for scoring. If the login is good, continues, otherwise it calls the callback with the error.
   * If there is a network outage, auth failure, or any other ip/tcp/http level error, it will log it and move on. It will NOT
   * prevent the login in this case.
   */
  public autoDecision (user: User, context: Context, callback: Callback, options: DecisionOptions = {}): void {
    try {
      const reqBody = this._buildBody(user, context, options);
      this._base.autoDecision(reqBody, options).then(() => {
        process.nextTick(() => callback(null, user, context));
      }, err => {
        process.nextTick(() => callback(err, user, context));
      });
    } catch (err) {
      this._logger.error(err);
      process.nextTick(() => callback(null, user, context));
    }
  }

  public async authFailure (user: User, context: Context, options: DecisionOptions = {}): Promise<CognitionResponse> {
    const response = await this.decision(user, context, {
      ...options,
      overrides: _.merge(options.overrides, {
        login: {
          status: LoginStatus.failure
        }
      })
    });

    return {
      decision: DecisionStatus.reject,
      signals: response.signals || ['failed_to_decision'],
      score: response.score || 0,
      confidence: response.confidence || 0,
      token: response.token || 'unknown'
    };
  }

  public static isGoodLogin (decisionResponse: CognitionResponse): boolean {
    return Login.isGoodLogin(decisionResponse);
  }

  private _getAuthenticationType (user: User, context: Context): AuthenticationType | undefined {
    const latestAuthMethod = _.last(_.sortBy(context.authentication.methods, 'timestamp'));

    if (typeof latestAuthMethod === 'undefined') {
      return undefined;
    } else if (latestAuthMethod.name === 'mfa') {
      return AuthenticationType.two_factor;
    } else if (latestAuthMethod.name === 'federated') {
      const identity = _.find(user.identities, { connection: context.connection });
      // check social VS sso
      if (typeof identity === 'undefined' || identity.isSocial) {
        return AuthenticationType.social_sign_on;
      } else {
        return AuthenticationType.single_sign_on;
      }
    } else if (_.get(context, 'sso.current_clients', []).length > 0) {
      return AuthenticationType.client_storage;
    } else {
      // Currently password-less still falls to password
      return AuthenticationType.password;
    }
  }

  private _buildBody (user: User, context: Context, options: DecisionOptions): CognitionInput {
    let request: CognitionInput = {
      _custom: {
        // Include Auth0 Specific data points
        auth0: {
          sdkVersion: '1.0',
          user: {
            updated: user.updated_at,
            fullName: user.name,
            lastName: user.family_name,
            firstName: user.given_name,
            username: user.username,
            email: user.email,
            emailVerified: user.email_verified || false,
            phoneNumber: user.phone_number,
            phoneNumberVerified: user.phone_verified || false,
            blocked: user.blocked || false
          },
          context: {
            authenticationMethods: context.authentication.methods,
            stats: context.stats,
            geoIp: context.request.geoip,
            primaryUser: context.primaryUser,
            ssoCurrentClients: context.sso.current_clients
          }
        }
      },
      eventId: _.get(context.request.query, 'cognition_event_id'),
      ipAddress: context.request.ip,
      login: {
        userId: user.user_id,
        channel: Channel.web,
        usedCaptcha: false,
        usedRememberMe: false,
        authenticationType: this._getAuthenticationType(user, context),
        status: LoginStatus.success,
        passwordUpdateTime: user.last_password_reset
      }
    };

    if (options.privacyMode) {
      request = _.omit(request, [
        'fullName',
        'lastName',
        'firstName',
        'email',
        'phoneNumber'
      ].map((field) => `_custom.auth0.user.${field}`)) as CognitionInput;
    }

    return _.merge(request, options.overrides);
  }
}

export { Auth0 };
export { HttpError, DecisionError } from './lib/error';
