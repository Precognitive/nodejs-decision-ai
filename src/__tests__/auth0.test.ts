import _ from 'lodash';
import fetchMock from '../__mocks__/node-fetch';
import { Auth0, DecisionError, HttpError } from '../auth0';
import { ApiVersion, CognitionResponse, DecisionStatus, AuthenticationType, Channel } from '../lib/decisionAi';
import { ContextProtocol, User, Context } from '../lib/auth0';
import fetchMockModule = require('fetch-mock');

const date = new Date();

// from https://stackoverflow.com/questions/3143070/javascript-regex-iso-datetime
const isoDateRegexp = /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/;

function getBaseResponse (decision: DecisionStatus): CognitionResponse {
  return {
    decision,
    score: _.random(-200, 200),
    confidence: _.random(0, 100),
    signals: [
      'test',
      'test2'
    ]
  }
}

const config = {
  auth: {
    userName: 'test',
    password: 'test-pw'
  },
  apiKey: 'kasf8w-afsafs-1asffw',
  version: ApiVersion.v1
};
const basicAuth = 'Basic dGVzdDp0ZXN0LXB3';

function getOptions (): {user: User, context: Context} {
  return {
    user: {
      name: 'test',
      app_metadata: {},
      email: "test@example.com",
      last_ip: "127.0.0.1",
      last_login: date,
      logins_count: 3,
      last_password_reset: date,
      password_set_date: date,
      created_at: date,
      updated_at: date,
      username: "test-example",
      user_id: "34294892831981",
      user_metadata: {}
    },
    context: {
      clientID: '123456789',
      sessionID: '123456789',
      protocol: ContextProtocol.OidcBasicProfile,
      request: {
        userAgent: 'test',
        ip: '127.0.0.1',
        hostname: 'example.com',
        query: '',
        geoip: {
          country_code: 'us',
          country_code3: '',
          country_name: 'murica',
          city_name: 'springville',
          latitude: '0.00',
          longitude: '0.00',
          time_zone: '-400',
          continent_code: 'na',
        }
      }
    }
  }
}

function getDecisionRecord () {
  return {
    apiKey: config.apiKey,
    dateTime: expect.stringMatching(isoDateRegexp),
    eventId: '123456789',
    ipAddress: '127.0.0.1',
    login: {
      authenticationType: 'password',
      channel: 'web',
      passwordUpdateTime: date.toISOString(),
      status: 'success',
      usedCaptcha: false,
      userId: '34294892831981',
    },
  };
}

function assertOneCall(calls: fetchMockModule.MockCall[], body: any) {
  expect(calls.length).toEqual(1);

  // the fetch mock api is asinine
  // it adds a 'request' property to the array that we don't care about
  // this strips the property and turns it into a real array.
  const request = [...calls[0]];

  expect(request).toEqual([url, {
    body: expect.any(String),
    headers: {
      Authorization: basicAuth
    },
    method: 'POST',
    timeout: 5000
  }]);
  expect(JSON.parse(request[1].body)).toEqual(body);
}

const url = 'https://api.precognitive.io/v1/decision/login';

afterEach(fetchMock.restore);

describe('decision', () => {
  it('returns decision-ai response', async () => {
    const defaultOptions = getOptions();
    const auth0 = new Auth0(config);
    const response = getBaseResponse(DecisionStatus.allow);
    fetchMock.postOnce(url, response);
    const actual = await auth0.decision(defaultOptions.user, defaultOptions.context);
    expect(actual).toEqual(response);
    assertOneCall(fetchMock.calls(), getDecisionRecord());
    expect(fetchMock.done()).toEqual(true);
  });

  it('throws error on http failure', async () => {
    const defaultOptions = getOptions();
    const auth0 = new Auth0(config);
    fetchMock.postOnce(url, {status: 500, body: JSON.stringify({code: 'failed', message: 'Failed'})});
    await expect(auth0.decision(defaultOptions.user, defaultOptions.context)).rejects.toBeInstanceOf(HttpError);
    expect(fetchMock.done()).toEqual(true);
  });

  it('throws error on request failure', async () => {
    const defaultOptions = getOptions();
    const auth0 = new Auth0(config);
    const err = new Error('Mock');
    fetchMock.postOnce(url, {throws: err});
    await expect(auth0.decision(defaultOptions.user, defaultOptions.context)).rejects.toBe(err);
    expect(fetchMock.done()).toEqual(true);
  });
});

describe('autodecision', () => {
  it('disallows if the response is `reject`', async () => {
    const defaultOptions = getOptions();
    const auth0 = new Auth0(config);
    const response = getBaseResponse(DecisionStatus.reject);
    fetchMock.postOnce(url, response);
    await expect(new Promise((resolve, reject) => {
      auth0.autoDecision(defaultOptions.user, defaultOptions.context, (err) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    })).rejects.toBeInstanceOf(DecisionError);
    assertOneCall(fetchMock.calls(), getDecisionRecord());
    expect(fetchMock.done()).toEqual(true);
  });

  it('allows if the response is `review`', async () => {
    const defaultOptions = getOptions();
    const auth0 = new Auth0(config);
    const response = getBaseResponse(DecisionStatus.review);
    fetchMock.postOnce(url, response);
    await new Promise((resolve, reject) => {
      auth0.autoDecision(defaultOptions.user, defaultOptions.context, (err) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
    assertOneCall(fetchMock.calls(), getDecisionRecord());
    expect(fetchMock.done()).toEqual(true);
  });

  it('allows if the response is `allow`', async () => {
    const defaultOptions = getOptions();
    const auth0 = new Auth0(config);
    const response = getBaseResponse(DecisionStatus.allow);
    fetchMock.postOnce(url, response);
    await new Promise((resolve, reject) => {
      auth0.autoDecision(defaultOptions.user, defaultOptions.context, (err) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
    assertOneCall(fetchMock.calls(), getDecisionRecord());
    expect(fetchMock.done()).toEqual(true);
  });

  it('allows on http failure', async () => {
    const defaultOptions = getOptions();
    const auth0 = new Auth0(config);
    fetchMock.postOnce(url, {status: 500, body: JSON.stringify({code: 'failed', message: 'Failed'})});
    await new Promise((resolve, reject) => {
      auth0.autoDecision(defaultOptions.user, defaultOptions.context, (err) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
    expect(fetchMock.done()).toEqual(true);
  });

  it('allows on request failure', async () => {
    const defaultOptions = getOptions();
    const auth0 = new Auth0(config);
    const err = new Error('Mock');
    fetchMock.postOnce(url, {throws: err});
    await new Promise((resolve, reject) => {
      auth0.autoDecision(defaultOptions.user, defaultOptions.context, (err) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
    expect(fetchMock.done()).toEqual(true);
  });

  it('allows on unexpected failure', async () => {
    const defaultOptions = getOptions();
    const auth0 = new Auth0(config);
    await new Promise((resolve, reject) => {
      auth0.autoDecision(defaultOptions.user, defaultOptions.context, (err) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      }, {
        overrides: {
          get login (): any {
            throw new Error('foo');
          }
        }
      });
    });
    // only assertion is code coverage
    expect(fetchMock.done()).toEqual(true);
  })
});

describe('context protocol => authentication type', () => {
  _.forEach([
    [ContextProtocol.OidcBasicProfile, AuthenticationType.password],
    [ContextProtocol.OidcImplicitProfile, AuthenticationType.password],
    [ContextProtocol.OAuth2ResourceOwner, AuthenticationType.password],
    [ContextProtocol.OAuth2Password, AuthenticationType.password],
    [ContextProtocol.SAMLP, AuthenticationType.single_sign_on],
    [ContextProtocol.WSFed, AuthenticationType.single_sign_on],
    [ContextProtocol.WSTrustUsernameMixed, AuthenticationType.single_sign_on],
    [ContextProtocol.OAuth2RefreshToken, AuthenticationType.key],
    [ContextProtocol.OAuth2ResourceOwnerJwtBearer, AuthenticationType.key],
    [ContextProtocol.Delegation, undefined],
    [ContextProtocol.RedirectCallback, undefined],
    ['invalid', undefined],
  ] as [ContextProtocol, AuthenticationType][], ([contextProtocol, authType]) => {
    it(`converts '${contextProtocol}' into '${authType}'`, async () => {
      const defaultOptions = getOptions();
      defaultOptions.context.protocol = contextProtocol;
      const auth0 = new Auth0(config);
      const response = getBaseResponse(DecisionStatus.allow);
      fetchMock.postOnce(url, response);
      const actual = await auth0.decision(defaultOptions.user, defaultOptions.context);
      expect(actual).toEqual(response);

      const expectedBody = getDecisionRecord();
      expectedBody.login.authenticationType = authType;
      assertOneCall(fetchMock.calls(), expectedBody);;
      expect(fetchMock.done()).toEqual(true);
    });
  });
});

describe('isGoodLogin', () => {
  it('accepts allowed login', () => {
    expect(Auth0.isGoodLogin(getBaseResponse(DecisionStatus.allow))).toEqual(true);
  });

  it('accepts reviewed login', () => {
    expect(Auth0.isGoodLogin(getBaseResponse(DecisionStatus.review))).toEqual(true);
  });

  it('rejects rejected login', () => {
    expect(Auth0.isGoodLogin(getBaseResponse(DecisionStatus.reject))).toEqual(false);
  });
});

it('allows field override', async () => {
  const defaultOptions = getOptions();
  const auth0 = new Auth0(config);
  const response = getBaseResponse(DecisionStatus.allow);
  fetchMock.postOnce(url, response);
  const actual = await auth0.decision(defaultOptions.user, defaultOptions.context, {
    overrides: {
      login: {
        channel: Channel.app
      }
    }
  });
  expect(actual).toEqual(response);

  const expectedBody = getDecisionRecord();
  expectedBody.login.channel = Channel.app;
  assertOneCall(fetchMock.calls(), expectedBody);
  expect(fetchMock.done()).toEqual(true);
});

it('uses custom logger', async () => {
  const logger = {
    debug: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn()
  };
  const defaultOptions = getOptions();
  const auth0 = new Auth0({...config, logger});
  const response = getBaseResponse(DecisionStatus.allow);
  fetchMock.postOnce(url, response);
  await auth0.decision(defaultOptions.user, defaultOptions.context);
  expect(_.mapValues(logger, (mock) => mock.mock.calls)).toMatchSnapshot();
  expect(fetchMock.done()).toEqual(true);
});
