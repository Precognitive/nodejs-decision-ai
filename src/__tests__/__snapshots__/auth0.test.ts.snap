// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uses custom logger 1`] = `
Object {
  "debug": Array [
    Array [
      "REQUEST BODY - {\\"dateTime\\":\\"2019-04-16T20:41:41.542Z\\",\\"eventId\\":\\"123456789\\",\\"ipAddress\\":\\"127.0.0.1\\",\\"login\\":{\\"userId\\":\\"34294892831981\\",\\"channel\\":\\"web\\",\\"usedCaptcha\\":false,\\"authenticationType\\":\\"password\\",\\"status\\":\\"success\\",\\"passwordUpdateTime\\":\\"2019-04-16T20:41:41.351Z\\"},\\"apiKey\\":\\"kasf8w-afsafs-1asffw\\"}",
    ],
    Array [
      "RESPONSE BODY - {\\"decision\\":\\"allow\\",\\"score\\":-8,\\"confidence\\":22,\\"signals\\":[\\"test\\",\\"test2\\"]}",
    ],
  ],
  "error": Array [],
  "info": Array [],
  "warn": Array [],
}
`;
